### PR TITLE
testgen: add getBlockReceipts testcase

### PR DESCRIPTION
**Pull request by jsvisa**

### CODE REVIEW

File: testgen/generators.go [hunk 1]
```diff
@@ -862,6 +863,76 @@ var EthGetTransactionReceipt = MethodTests{
...
 }
 
+var EthGetBlockReceipts = MethodTests{
+	"eth_getBlockReceipts",
+	[]Test{
+		{
+			"get-block-0",
+			"gets receipts for block 0",
+			func(ctx context.Context, t *T) error {
+				return t.rpc.CallContext(ctx, nil, "eth_getBlockReceipts", "0x0")
```
Comment (lightclient) [inline testgen/generators.go:873]: Can we verify the result against the expected value? I do this on all the test to guard against the filling client misinterpreting the query and spitting out the wrong receipts. The way I usually do this is access the actual `Blockchain` object in `T` (`t.chain`) and pull the values from the structure to compare against the RPC result.

The tests you wrote for `eth_getTransactionReceipt` did this correctly.

**Review by lightclient — CHANGES_REQUESTED** [review]

File: testgen/generators.go (hunk 1 above)
Comment (jsvisa) [reply testgen/generators.go:873]: 🙆‍♂️

Comment (jsvisa) [reply testgen/generators.go]: please take another look, thanks :)

**Comment by jsvisa** [comment]
BTW, I'm not using `rpc.BlockNumberOrHash` to replace `number, hash, tag`, so I'm trying to simulate the real client requests(eg: web3.js, web3.py) in order to minimize coupling with geth

**Comment by jsvisa** [comment]
BTW, I'm not using `rpc.BlockNumberOrHash` to replace `number, hash, tag`, so I'm trying to simulate the real client requests(eg: web3.js, web3.py) in order to minimize coupling with geth

File: testgen/generators.go [hunk 2]
```diff
@@ -862,6 +863,140 @@ var EthGetTransactionReceipt = MethodTests{
...
+		if !bytes.Equal(got, want) {
+			return fmt.Errorf("receipt mismatch (got: %s, want: %s)", hexutil.Bytes(got), hexutil.Bytes(want))
+		}
+	}
+	return nil
+}
+
+func testBlockReceiptsByNumber(ctx context.Context, t *T, number uint64) error {
+	var receipts []*types.Receipt
+	if err := t.rpc.CallContext(ctx, &receipts, "eth_getBlockReceipts", fmt.Sprintf("0x%x", number)); err != nil {
```
Comment (lightclient) [inline testgen/generators.go:883]:
```suggestion
	if err := t.rpc.CallContext(ctx, &receipts, "eth_getBlockReceipts", hexutil.Uint(number)); err != nil {
```

File: testgen/generators.go [hunk 3]
```diff
@@ -862,6 +863,140 @@ var EthGetTransactionReceipt = MethodTests{
...
+func checkBlockReceipts(t *T, hash common.Hash, receipts []*types.Receipt) error {
+	blockReceipts := t.chain.GetReceiptsByHash(hash)
+	if len(receipts) != len(blockReceipts) {
+		return fmt.Errorf("receipts length notmatch(got: %d, want: %d)", len(receipts), len(blockReceipts))
+	}
+	for i := range receipts {
+		got, _ := receipts[i].MarshalBinary()
+		want, _ := blockReceipts[i].MarshalBinary()
+		if !bytes.Equal(got, want) {
+			return fmt.Errorf("receipt mismatch (got: %s, want: %s)", hexutil.Bytes(got), hexutil.Bytes(want))
```
Comment (lightclient) [inline testgen/generators.go:875]:
```suggestion
			return fmt.Errorf("receipt mismatch (got: %x, want: %x)", got, want)
```

File: testgen/generators.go [hunk 4]
```diff
@@ -862,6 +863,140 @@ var EthGetTransactionReceipt = MethodTests{
...
+	var receipts []*types.Receipt
+	if err := t.rpc.CallContext(ctx, &receipts, "eth_getBlockReceipts", tag); err != nil {
+		return err
+	}
+	var hash common.Hash
+	switch tag {
+	case "earliest":
+		block := t.chain.GetBlockByNumber(0)
+		hash = block.Hash()
+	case "latest":
```
Comment (lightclient) [inline testgen/generators.go:913]: Should have `finalized` case.

File: testgen/generators.go [hunk 5]
```diff
@@ -862,6 +863,140 @@ var EthGetTransactionReceipt = MethodTests{
...
+			"get-block-future",
+			"gets receipts of future block",
+			func(ctx context.Context, t *T) error {
+				block := t.chain.CurrentBlock()
+				if block == nil {
+					return errors.New("current block notfound")
+				}
+				number := block.Number.Int64()
+				var receipts []*types.Receipt
+				if err := t.rpc.CallContext(ctx, &receipts, "eth_getBlockReceipts", fmt.Sprintf("0x%x", number+1)); err != nil {
```
Comment (lightclient) [inline testgen/generators.go:949]:
```suggestion
				if err := t.rpc.CallContext(ctx, &receipts, "eth_getBlockReceipts", hexutil.Uint(number+1)); err != nil {
```

File: testgen/generators.go (hunk 4 above)
Comment (jsvisa) [reply testgen/generators.go:913]: AFAIK, `safe` and `finalized` are only available after the merge, and at least need 1,2 epochs. If those tags are available, we need to generate at least 64 blocks?

Comment (lightclient) [reply testgen/generators.go:913]: You're right. It looks like there is an issue with how merge status is determined in geth. Even if `TerminalTotalDifficulty` is set on genesis, it only _actually_ triggers if it receives a message over the engine api. Let me look into this a bit. Even so, it isn't well defined whether explicitly imported blocks (like what happens here with `geth import`) should be considered `safe`, `finalized`, etc.

Comment (jsvisa) [reply testgen/generators.go:913]: Let me take a look too :)

Comment (jsvisa) [reply testgen/generators.go:913]: I think we can run geth with jwt secret, and then post a `ForkchoiceUpdatedV2` message to notify the EL to post merge

**Review by lightclient — APPROVED** [review]
Hey thanks for moving that `final`, `safe` stuff to another PR. This one looks good. I moved the receipt checker to `utils.go` and simplified the test calls a bit. If possible, I prefer to write everything in the test, even if it means a bit of duplication as I find it clearer and easier to understand what's going on if I don't need to jump into a different file. Since 1 line vs. 3ish I think this is an acceptable tradeoff. Thanks for your work on this!

**Pull request merged**